### ISSUE TITLE
Display write time instead of sent time

### ISSUE THREAD

**Issue by breversa** [kontalk/androidclient#627]
Hi !

I noticed the following behaviour :
* Alice is away deep in the jungle (or anywhere else with no network). She writes Bob a message at time t.
* At t+10, she writes another message (still no network).
* At t+200 (yes, that was a long trip !), she finally gets network connection.
* Bob (who has activated "display message sent time") receives both messages (written at t and t+20), but the displayed sent time is t+200 for both.

-> I believe it would be more logical that each message displays its *write* time instead/in addition to its *send* time.

**Comment by daniele-athome** [comment]
Sorry that can't work because from the point of view of the server the message was sent at t+200. Since the recipient shows that time, it makes sense to match the same information in the sender.

**Comment by daniele-athome** [comment]
Besides, how would you justify this different time to the recipient? :-)

**Comment by abika** [comment]
well, from the technically side you could add a "compose time" to a message when sending it (dunno if there is an XEP for that). And the receiver could see three timestamps now in the "message details":
* compose time
* sent time
* received time

However, this makes it more confusing and can be a privacy problem: there are maybe good reasons for not exposing the time you wrote a message.

**Comment by breversa** [comment]
@abika :
Yes, that's what I thought : adding an extra info to the message itself.

However, I don't quite see how/why you (as the sender) would NOT want (the receiver) to know the compose time, but have no problem with the sent time ? :-)

And considering there was a sound reason, why not offer an option (to the sender) to provide the compose time that the receiver could display if present ?
(Of course, that may require a bit of explanation, but hey, I'm pretty sure Kontalk users are smart enough to understand) ;-)

**Comment by daniele-athome** [comment]
> Of course, that may require a bit of explanation, but hey, I'm pretty sure Kontalk users are smart enough to understand

You overestimate the average user :P

As for the protocol issue, I think `<delay/>` could be used for that. I don't think the server strips that (it adds it if it's not present but it should keep it if it finds it IIRC) - now that I think about it, it can also be a security leak (i.e. timestamp could be forged).

Still, I'm not that convinced. Servers exist also to mediate, not only just to route. The server guarantees the delivery timestamp (eeehhrr forget the previous paragraph, I'm going to fix that :-) and the client could send any forged information to make the recipient believe it was sent at another time than the servers declares. Besides, I'd look at the actual sent time in a message as a proof, not the time-when-it-was-supposedly-written.

Also, you can't tell someone you tried to call him/her and prove it if you didn't have network coverage at the time: you would have to delay the call until you had reached a covered area (sorry for the possible wrong use of the tenses here). The same restriction should exist for messages too IMO: if you can't contact someone, you just can't; you can't fake to have done so in a different time (and I'm afraid people would take advantage of just that).

Summary: sorry I'm not approving this.

**Comment by abika** [comment]
@breversa 
> However, I don't quite see how/why you (as the sender) would NOT want (the receiver) to know the compose time, but have no problem with the sent time ? :-)

By principle composing a message is only done by the client, so you can totally hide that from anybody else. Sending a message is different: the server is obviously involved and if it decides to save the sent time the sender can't do anything against it.
And a (fabricated) example: I intentionally disconnect the client and write a long message to my superior. A long time after that connect to send the message, my superior does not (and should not) know when I wrote the message.

Yes, the "best" solution would be having three timestamps + an option to disable the "compose timestamp". But as @daniele-athome pointed out this doesn't solve the problem of faking the "compose timestamp", so this is not a "good" solution.

TL;DR: agree with `wontfix`

**Comment by breversa** [comment]
Hi, and thanks for your input regarding this issue.

Though I understand the potential to forge compose timestamps, I still believe it could be a useful feature to some people (starting with me, I won't hide it !). What is needed (IMO) is a way to make users aware that a compose timestamp is NOT foolproof. If compose timestamp is optional, why not display a warning banner (in the same style as when the user key has changed) stating that "Compose time is merely a courtesy of the sender and NOT a proof that the message was really composed at that time" (or something similar) ? The keyword here is "courtesy". :-)

Alternatively, I thought a bit more about it and imagined the following : instead of an option to "disable compose time", why no do the contrary and add an option to "add compose time" to outgoing message, IRC-style, like this :

Alice :
[16:40 GMT] Hey, how are you ?

Bob :
[16:41 GMT] I'm fine, thanks !

The "[16:40 GMT]" would be a simple prefix to the outgoing message. Not metadata, nothing fancy, just… a prefix.

**Comment by daniele-athome** [comment]
IMO a solution like that will increase the complexity for a not so big advantage in an app that wants to be as simple as humanly possible.
I'm sorry I really can't find the usefulness in the very concept of a compose timestamp. On the contrary, it introduces confusion due to yet another date/time to be considered, for no real advantage.

**Comment by abika** [comment]
for the sake of completeness: I agree with Daniele

**Comment by breversa** [comment]
@daniele-athome :
Is Kontalk supposed to be "as simple as humanely possible", or provide "as many useful features as possible" (though I understand that all features may not appeal to everyone) ? That's an open question, of course. You know from our e-mail exchanges that I only wish the best (= widest-spread) for Kontalk. :-)

The simplicity aspect is precisely why I suggested the compose timestamp to be :
1- optional and not default, so that it cannot confuse lambda users that see no interest in it.
2- a simple prefix, so that it doesn't add much complexity, code-wise.

(Also, the compose timestamp could help spot/debug connexion problems with the server. But that's just a side effect and no the main use, so I did not put it forward.)

Outside the extra complexity (which I assume could be rather low) and little advantage (which I understand is a bigger thing for me than others), is there any other objection to this suggestion ? I do not intend to force this feature down anyone's throat (how could I anyway) ; just trying to figure out solutions for problems that I may not spontaneously see.
